zeus/utils/framework.py

```diff
@@ -102,3 +102,53 @@ def sync_execution(
         return
 
     raise RuntimeError("No framework is available.")
+
+
+def all_reduce(
+    object: List[int] | List[float], operation: Literal["sum", "max"]
+) -> int | float:
+    """Reduce objects from all replicas through the specified operation.
+
+    If running in a distributed setting, the objects are reduced across all replicas.
+    If running in a non-distributed setting, the operation is just done on the single object.
```

Comment:
No, this deviates from the definition of AllReduce. In a distributed setting, numbers in the input tensor should be reduced *elementwise*. In a non-distributed setting, `object` should be returned as is. See also the [official definition of collective communication by NCCL](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html) -- You can see in the AllReduce section that the sizes of tensors (rectangles) are identical across all nodes are identical.

```
Rank  Input       => Output (reduce op is sum)
0     [0.0, 1.0]  => [2.0, 4.0]
1     [2.0, 3.0]  => [2.0, 4.0]
```

This didn't get noticed because you were calling `all_reduce` on lists with only one element, but what you implemented is not AllReduce and it's going to cause a lot of confusion down the line for future contributors. Please revise the implementation to match the definition of the AllReduce collective.

Reply:
Ah I see. I'll fix that!